firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+        let imageView = loadGifFromBundle(named: "easterEggGif")
+        imageView?.isHidden = true
+        imageView?.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
+    private var isIpad: Bool {
+        traitCollection.userInterfaceIdiom == .pad && traitCollection.horizontalSizeClass == .regular
+    }
+
+    init(parentScrollView: UIScrollView?,
+         isPotraitOrientation: Bool,
+         onRefreshCallback: @escaping VoidReturnCallback) {
+        self.scrollView = parentScrollView
+        self.onRefreshCallback = onRefreshCallback
+        super.init(frame: .zero)
+        clipsToBounds = true
+        setupEasterEgg(isPotrait: isPotraitOrientation)
+        setupSubviews()
+        startObservingContentScroll()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private func setupEasterEgg(isPotrait: Bool) {
+        if let easterEggGif {
+            addSubview(easterEggGif)
+            let shrinkFactor = computeShrinkingFactor()
+            let easterEggSize = UX.easterEggSize.applying(CGAffineTransform(scaleX: shrinkFactor, y: shrinkFactor))
+            let layoutBuilder = EasterEggViewLayoutBuilder(easterEggSize: easterEggSize)
+            layoutBuilder.layoutEasterEggView(easterEggGif, superview: self, isPotrait: isPotrait, isIpad: isIpad)
+        }
+    }
+
+    private func setupSubviews() {
+        addSubviews(progressContainerView, progressView)
+        let shrinkFactor = computeShrinkingFactor()
+        let progressContainerViewPadding = UX.progressViewPadding * shrinkFactor
+        let progressContainerViewSize = UX.progressViewAnimatedBackgroundSize * shrinkFactor
+        NSLayoutConstraint.activate([
+            progressContainerView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -progressContainerViewPadding),
+            progressContainerView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            progressContainerView.heightAnchor.constraint(equalToConstant: progressContainerViewSize),
+            progressContainerView.widthAnchor.constraint(equalToConstant: progressContainerViewSize),
+
+            progressView.centerYAnchor.constraint(equalTo: progressContainerView.centerYAnchor),
+            progressView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            progressView.heightAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor),
+            progressView.widthAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor)
+        ])
+    }
+
+    /// Starts observing the `UIScrollView` content offset so the refresh can react to content change and show
+    /// the correct state of the pull to refresh view.
+    func startObservingContentScroll() {
+        obeserveTicket = scrollView?.observe(\.contentOffset) { [weak self] _, _ in
+            guard let scrollView = self?.scrollView, scrollView.isDragging
+            else {
```

Comment:
```suggestion
            guard let scrollView = self?.scrollView, scrollView.isDragging else {
```
(I think we like to put the else on the same line unless each guard check is on its own line)